format/ecs.js

```diff
@@ -0,0 +1,29 @@
+/* eslint-disable no-param-reassign */
+const { format } = require('logform');
+
+/*
+ * function ecs (info)
+ * Aligns the base info object fields to the ECS schema
+ */
+module.exports = format((info) => {
```

Comment:
I'm not sure I understand the implementation of this formatter. 
It relates and filters some of the fields which could be presented in the `info` object, what if there are other fields? We just filter them out when we're picking the json?
My point is, maybe instead we want instead to define the format that we're aiming to?
From here I can't really understand how does the final log format looks like.
Also, `delete` in javascript is a bit costly operation. It's recommended to create a new object and omit the properties that we don't want there.

Reply:
How I understand it: winston has essentially two types of "formatters". These that mutate the `info` and "finalisers". I.e a formatter that supposed to be the serilizer in the end on the formatter pipeline. For example, `console` or `json` are finalisers, while `timestamp` or `metadata` are not.

In this case I made ecs to be non-finaliser.  It's only goal is to tweak field labels that come from the rest of the pipeline.

Reply:
This is, by the way, a very common pattern in logging formatters in other languages as well (namely Python/Go).

Reply:
>It's highly unlikely any other fields would be added to the base without our knowledge.

That's why this implementation feels strange for me. Probably, I'm missing something.
Or we assume that only our logger (+ formatters used) changes the base fields and then we don't need all of the `ifs`, because we know how the `info` looks like, or we assume that it can change, then we should use a clear explicit format instead of addressing specific fields.

Reply:
OK,  I think I understand better now.

`timestamp` is added by a previous formatter. The `if` can be removed if `timestamp` formatter is always there before `ecs`. `level` should be always there, probably the only one that needs no guarding.
`stack` and `strigo` are optional and needs to be guarded.

Reply:
@v0vs official examples from winston re `delete`: https://github.com/winstonjs/winston/blob/master/examples/delete-level.js

wdyt?

Reply:
The Winston docs say you can currently treat the object as mutable, but I think it might be cleaner to just construct a new object and explicitly pass the logged fields. But that's just me and my strong typing fetish

Reply:
Just to make it clear - delete on itself won't hurt that much, despite that the assigning undefined to a property is 40 times faster than delete, but we're talking about a difference in several ms, which is not really important.
Deleting one property does not bother me, but when I see several properties deleted in a row that raises a question about the implementation design.
By default I'm against the mutable objects as much as we can, and what @holograph said because it does not introduce any side effects.

Reply:
I sincerely hope you meant several µs (or even ns), not ms, otherwise we really are screwed :-)
The benefit of constructing an object is, for my money, in documenting the expected logger output; side benefits may include performance or future-proofing wins, but that's not really why it's important.

Reply:
I think winston provides a simple profiler that we can use to check.. I'll try it tomorrow.

Reply:
> Deleting one property does not bother me, but when I see several properties deleted in a row that raises a question about the implementation design.
By default I'm against the mutating some global/arguments objects as much as we can, and what @holograph said because it does not introduce any side effects.

Can someone show alternative approach? Info contains Symbols as well. What is the quickest way to "deep copy an object, including its symbols, excluding extra fields"

Reply:
Just doing a very quick and dirty check like this:

```

const { setupNodeLogger } = require('../setup/winston');

// Initiate a logger with 'info' level in JSON format
const log = setupNodeLogger({});

let then = new Date;
for (let i = 0; i < 1000; i++) {
  log.debug("profiling", {workspaceId: "blah"});
}
let now = new Date;


console.log(now - then);
```

shows no obvious speedups

Reply:
iv'e updated the formatter to include less logic, I think you can be happy with it now

Reply:
Kind of this
```
  const {
    timestamp, level, log, ...filteredInfo
  } = info;

  const newInfo = { '@timestamp': timestamp, log: { level, ...log }, ...filteredInfo };

  return newInfo;
```
The first row just extracts `timestamp`, `level` and `log` and leaves the left properties in `filteredInfo`, which is a new shallow copied object.

Reply:
what about symbols?

Reply:
Copies symbols as well

Reply:
it does indeed.

I've just checked it with the 10k logs loop, and it looks bit more slower than the `delete` version.